pkg/cachemanager/cachemanager.go

```diff
@@ -0,0 +1,385 @@
+	}
+
+	if isNamespaceExcluded {
+		c.tracker.ForData(instance.GroupVersionKind()).CancelExpect(instance)
+		return nil
+	}
+
+	syncKey := syncutil.GetKeyForSyncMetrics(instance.GetNamespace(), instance.GetName())
+	if c.watchesGVK(gvk) {
+		_, err = c.cfClient.AddData(ctx, instance)
+		if err != nil {
+			c.syncMetricsCache.AddObject(
+				syncKey,
+				syncutil.Tags{
+					Kind:   instance.GetKind(),
+					Status: metrics.ErrorStatus,
+				},
+			)
+
+			return err
+		}
+
+		c.syncMetricsCache.AddObject(syncKey, syncutil.Tags{
+			Kind:   instance.GetKind(),
+			Status: metrics.ActiveStatus,
+		})
+		c.syncMetricsCache.AddKind(instance.GetKind())
+	}
+
+	c.tracker.ForData(instance.GroupVersionKind()).Observe(instance)
+
+	return nil
+}
+
+func (c *CacheManager) RemoveObject(ctx context.Context, instance *unstructured.Unstructured) error {
+	gvk := instance.GroupVersionKind()
+
+	if c.watchesGVK(gvk) {
+		if _, err := c.cfClient.RemoveData(ctx, instance); err != nil {
+			return err
+		}
+	}
+
+	// only delete from metrics map if the data removal was successful
+	c.syncMetricsCache.DeleteObject(syncutil.GetKeyForSyncMetrics(instance.GetNamespace(), instance.GetName()))
+	c.tracker.ForData(instance.GroupVersionKind()).CancelExpect(instance)
+
+	return nil
+}
+
+func (c *CacheManager) wipeData(ctx context.Context) error {
+	if _, err := c.cfClient.RemoveData(ctx, target.WipeData()); err != nil {
+		return err
+	}
+
+	// reset sync cache before sending the metric
+	c.syncMetricsCache.ResetCache()
+	c.syncMetricsCache.ReportSync()
+
+	return nil
+}
+
+func (c *CacheManager) ReportSyncMetrics() {
+	c.syncMetricsCache.ReportSync()
+}
+
+func (c *CacheManager) syncGVK(ctx context.Context, gvk schema.GroupVersionKind) error {
+	u := &unstructured.UnstructuredList{}
+	u.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   gvk.Group,
+		Version: gvk.Version,
+		Kind:    gvk.Kind + "List",
+	})
+
+	var err error
+	func() {
+		c.mu.Lock()
+		defer c.mu.Unlock()
+
+		// only call List if we are still watching the gvk.
+		if c.watchedSet.Contains(gvk) {
+			err = c.reader.List(ctx, u)
+		}
+	}()
+
+	if err != nil {
+		return fmt.Errorf("replaying data for %+v: %w", gvk, err)
+	}
+
+	for i := range u.Items {
+		if err := c.AddObject(ctx, &u.Items[i]); err != nil {
+			return fmt.Errorf("adding data for %+v: %w", gvk, err)
+		}
+	}
+
+	return nil
+}
+
+func (c *CacheManager) manageCache(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-c.backgroundManagementTicker.C:
+			func() {
+				c.mu.Lock()
+				defer c.mu.Unlock()
+
+				c.wipeCacheIfNeeded(ctx)
+
+				// spin up new goroutines to relist gvks as there has been a wipe
+				if c.needToList {
+					// stop any goroutines that were relisting before
+					// as we may no longer be interested in those gvks
+					close(c.relistStopChan)
+
+					// assume all gvks need to be relisted
+					gvksToRelist := c.gvksToSync.GVKs()
```

Comment:
It seems that we are accessing a property on our struct using a receiver method (`GVKs()`) then passing that into another receiver method (`replayGVKs`) as a parameter.  I'm left asking "why can't we just access this in the other receiver method?  If we want to make a copy (which we do in `replayGVKs`), why don't we just have the accessor method (`GVKs`) return a copy and call it in `replayGVKs()`?

Reply:
After review I see that `GVKs()` returns a slice of GroupVersionKind objects and not a `watch.Set`.

Reply:
Curious for your thoughts here @maxsmythe

Reply:
Also, per our discussion, in this block here we have a lock in hand when we make this copy and work on this in the separate goroutine as opposed to reading the value in the goroutine under lock.

Reply:
Calling this to create a snapshot makes sense. Otherwise, there would be race conditions.

Also, b/c `replayGVKs()` is run as a separate goroutine, having it grab the list of GVKs to replay would complicate the locking logic.

Reply:
Closing this for now since we are all aligned.